Command line argument validation returns incorrect process exit code

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  3.6.3, 3.6.2 and typescript@next

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:**
build option file not found
**Code**
```shell
tsc -b bogus.json
echo $?
```

**Expected behavior:**
In 3.5.3, this still prints the following:
```
> tsc -b bogus.json
< error TS6053: File 'bogus.jon/tsconfig.json' not found.
> echo $?
< 1
```

**Actual behavior:**
Since 3.6.2, this prints the following:
```
> tsc -b bogus.json
< error TS6053: File 'bogus.jon/tsconfig.json' not found.
> echo $?
< 0
```

As such, it is not possible to detect for infastructure scripts (that invoke `tsc`) to know when `tsc` was not able to find the `tsconfig.json` file.

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior -->

**Related Issues:** <!-- Did you find other bugs that looked similar? -->

The same issue happens for the following error:
```
error TS18003: No inputs were found in config file 'tsconfig.json'. Specified 'include' paths were '["**/*"]' and 'exclude' paths were '[]'.
```
Therefore I assume that all process errors from the command line args validation is returning an error code of 0 instead of 1.

--- a/src/compiler/tsbuild.ts
+++ b/src/compiler/tsbuild.ts
@@ -1807,7 +1807,7 @@ namespace ts {
         startWatching(state, buildOrder);
 
         return isCircularBuildOrder(buildOrder) ?
-            ExitStatus.ProjectReferenceCycle_OutputsSkupped :
+            ExitStatus.ProjectReferenceCycle_OutputsSkipped :
             errorProjects ?
                 successfulProjects ?
                     ExitStatus.DiagnosticsPresent_OutputsGenerated :
@@ -1821,7 +1821,7 @@ namespace ts {
 
         if (isCircularBuildOrder(buildOrder)) {
             reportErrors(state, buildOrder.circularDiagnostics);
-            return ExitStatus.ProjectReferenceCycle_OutputsSkupped;
+            return ExitStatus.ProjectReferenceCycle_OutputsSkipped;
         }
 
         const { options, host } = state;

--- a/src/compiler/tsbuild.ts
+++ b/src/compiler/tsbuild.ts
@@ -1788,31 +1788,25 @@ namespace ts {
 
         let reportQueue = true;
         let successfulProjects = 0;
-        let errorProjects = 0;
         while (true) {
             const invalidatedProject = getNextInvalidatedProject(state, buildOrder, reportQueue);
             if (!invalidatedProject) break;
             reportQueue = false;
             invalidatedProject.done(cancellationToken);
-            if (state.diagnostics.has(invalidatedProject.projectPath)) {
-                errorProjects++;
-            }
-            else {
-                successfulProjects++;
-            }
+            if (!state.diagnostics.has(invalidatedProject.projectPath)) successfulProjects++;
         }
 
         disableCache(state);
         reportErrorSummary(state, buildOrder);
         startWatching(state, buildOrder);
 
-        return isCircularBuildOrder(buildOrder) ?
-            ExitStatus.ProjectReferenceCycle_OutputsSkipped :
-            errorProjects ?
-                successfulProjects ?
-                    ExitStatus.DiagnosticsPresent_OutputsGenerated :
-                    ExitStatus.DiagnosticsPresent_OutputsSkipped :
-                ExitStatus.Success;
+        return isCircularBuildOrder(buildOrder)
+            ? ExitStatus.ProjectReferenceCycle_OutputsSkipped
+            : !buildOrder.some(p => state.diagnostics.has(toResolvedConfigFilePath(state, p)))
+                ? ExitStatus.Success
+                : successfulProjects
+                    ? ExitStatus.DiagnosticsPresent_OutputsGenerated
+                    : ExitStatus.DiagnosticsPresent_OutputsSkipped;
     }
 
     function clean(state: SolutionBuilderState, project?: string, onlyReferences?: boolean) {

--- a/src/compiler/types.ts
+++ b/src/compiler/types.ts
@@ -3259,6 +3259,9 @@ namespace ts {
         InvalidProject_OutputsSkipped = 3,
 
         // When build is skipped because project references form cycle
+        ProjectReferenceCycle_OutputsSkipped = 4,
+
+        /** @deprecated Use ProjectReferenceCycle_OutputsSkipped instead. */
         ProjectReferenceCycle_OutputsSkupped = 4,
     }
 

--- a/src/testRunner/tsconfig.json
+++ b/src/testRunner/tsconfig.json
@@ -99,6 +99,7 @@
         "unittests/tsbuild/demo.ts",
         "unittests/tsbuild/emitDeclarationOnly.ts",
         "unittests/tsbuild/emptyFiles.ts",
+        "unittests/tsbuild/exitCodeOnBogusFile.ts",
         "unittests/tsbuild/graphOrdering.ts",
         "unittests/tsbuild/inferredTypeFromTransitiveModule.ts",
         "unittests/tsbuild/javascriptProjectEmit.ts",

--- a/src/testRunner/unittests/tsbuild/demo.ts
+++ b/src/testRunner/unittests/tsbuild/demo.ts
@@ -88,7 +88,7 @@ namespace ts {
     }
   ]`
                 ),
-                expectedExitStatus: ExitStatus.ProjectReferenceCycle_OutputsSkupped,
+                expectedExitStatus: ExitStatus.ProjectReferenceCycle_OutputsSkipped,
                 expectedDiagnostics: () => [
                     getExpectedDiagnosticForProjectsInBuild("src/animals/tsconfig.json", "src/zoo/tsconfig.json", "src/core/tsconfig.json", "src/tsconfig.json"),
                     errorDiagnostic([

--- a/src/testRunner/unittests/tsbuild/demo.ts
+++ b/src/testRunner/unittests/tsbuild/demo.ts
@@ -39,17 +39,19 @@ namespace ts {
 
         interface VerifyBuild {
             modifyDiskLayout: (fs: vfs.FileSystem) => void;
+            tsconfigs?: readonly string[];
             expectedExitStatus: ExitStatus;
             expectedDiagnostics: (fs: vfs.FileSystem) => fakes.ExpectedDiagnostic[];
             expectedOutputs: readonly string[];
             notExpectedOutputs: readonly string[];
         }
 
-        function verifyBuild({ modifyDiskLayout, expectedExitStatus, expectedDiagnostics, expectedOutputs, notExpectedOutputs }: VerifyBuild) {
+        function verifyBuild({ modifyDiskLayout, tsconfigs = ["/src/tsconfig.json"],
+                               expectedExitStatus, expectedDiagnostics, expectedOutputs, notExpectedOutputs }: VerifyBuild) {
             const fs = projFs.shadow();
             const host = fakes.SolutionBuilderHost.create(fs);
             modifyDiskLayout(fs);
-            const builder = createSolutionBuilder(host, ["/src/tsconfig.json"], { verbose: true });
+            const builder = createSolutionBuilder(host, tsconfigs, { verbose: true });
             const exitStatus = builder.build();
             assert.equal(exitStatus, expectedExitStatus);
             host.assertDiagnosticMessages(...expectedDiagnostics(fs));
@@ -151,6 +153,20 @@ namespace ts {
                     [Diagnostics.Skipping_build_of_project_0_because_its_dependency_1_has_errors, "/src/animals/tsconfig.json", "/src/core"],
                     [Diagnostics.Project_0_can_t_be_built_because_its_dependency_1_was_not_built, "src/zoo/tsconfig.json", "src/animals"],
                     [Diagnostics.Skipping_build_of_project_0_because_its_dependency_1_was_not_built, "/src/zoo/tsconfig.json", "/src/animals"],
+                ],
+                expectedOutputs: emptyArray,
+                notExpectedOutputs: [...coreOutputs(), ...animalOutputs(), ...zooOutputs()]
+            });
+        });
+
+        it("returns an error exit in case of a bogus file", () => {
+            verifyBuild({
+                modifyDiskLayout: noop,
+                tsconfigs: ["bogus.json"],
+                expectedExitStatus: ExitStatus.DiagnosticsPresent_OutputsSkipped,
+                expectedDiagnostics: _ => [
+                    getExpectedDiagnosticForProjectsInBuild("bogus.json"),
+                    errorDiagnostic([Diagnostics.File_0_not_found, "/bogus.json"]),
                 ],
                 expectedOutputs: emptyArray,
                 notExpectedOutputs: [...coreOutputs(), ...animalOutputs(), ...zooOutputs()]

--- a/src/testRunner/unittests/tsbuild/demo.ts
+++ b/src/testRunner/unittests/tsbuild/demo.ts
@@ -39,19 +39,17 @@ namespace ts {
 
         interface VerifyBuild {
             modifyDiskLayout: (fs: vfs.FileSystem) => void;
-            tsconfigs?: readonly string[];
             expectedExitStatus: ExitStatus;
             expectedDiagnostics: (fs: vfs.FileSystem) => fakes.ExpectedDiagnostic[];
             expectedOutputs: readonly string[];
             notExpectedOutputs: readonly string[];
         }
 
-        function verifyBuild({ modifyDiskLayout, tsconfigs = ["/src/tsconfig.json"],
-                               expectedExitStatus, expectedDiagnostics, expectedOutputs, notExpectedOutputs }: VerifyBuild) {
+        function verifyBuild({ modifyDiskLayout, expectedExitStatus, expectedDiagnostics, expectedOutputs, notExpectedOutputs }: VerifyBuild) {
             const fs = projFs.shadow();
             const host = fakes.SolutionBuilderHost.create(fs);
             modifyDiskLayout(fs);
-            const builder = createSolutionBuilder(host, tsconfigs, { verbose: true });
+            const builder = createSolutionBuilder(host, ["/src/tsconfig.json"], { verbose: true });
             const exitStatus = builder.build();
             assert.equal(exitStatus, expectedExitStatus);
             host.assertDiagnosticMessages(...expectedDiagnostics(fs));
@@ -153,20 +151,6 @@ namespace ts {
                     [Diagnostics.Skipping_build_of_project_0_because_its_dependency_1_has_errors, "/src/animals/tsconfig.json", "/src/core"],
                     [Diagnostics.Project_0_can_t_be_built_because_its_dependency_1_was_not_built, "src/zoo/tsconfig.json", "src/animals"],
                     [Diagnostics.Skipping_build_of_project_0_because_its_dependency_1_was_not_built, "/src/zoo/tsconfig.json", "/src/animals"],
-                ],
-                expectedOutputs: emptyArray,
-                notExpectedOutputs: [...coreOutputs(), ...animalOutputs(), ...zooOutputs()]
-            });
-        });
-
-        it("returns an error exit in case of a bogus file", () => {
-            verifyBuild({
-                modifyDiskLayout: noop,
-                tsconfigs: ["bogus.json"],
-                expectedExitStatus: ExitStatus.DiagnosticsPresent_OutputsSkipped,
-                expectedDiagnostics: _ => [
-                    getExpectedDiagnosticForProjectsInBuild("bogus.json"),
-                    errorDiagnostic([Diagnostics.File_0_not_found, "/bogus.json"]),
                 ],
                 expectedOutputs: emptyArray,
                 notExpectedOutputs: [...coreOutputs(), ...animalOutputs(), ...zooOutputs()]

--- a/src/testRunner/unittests/tsbuild/exitCodeOnBogusFile.ts
+++ b/src/testRunner/unittests/tsbuild/exitCodeOnBogusFile.ts
@@ -1,0 +1,11 @@
+namespace ts {
+    // https://github.com/microsoft/TypeScript/issues/33849
+    describe("unittests:: tsbuild:: exitCodeOnBogusFile:: test exit code", () => {
+        verifyTsc({
+            scenario: "exitCodeOnBogusFile",
+            subScenario: `test exit code`,
+            fs: () => loadProjectFromFiles({}, symbolLibContent),
+            commandLineArgs: ["-b", "bogus.json"]
+        });
+    });
+}

--- a/tests/baselines/reference/api/tsserverlibrary.d.ts
+++ b/tests/baselines/reference/api/tsserverlibrary.d.ts
@@ -1964,6 +1964,8 @@ declare namespace ts {
         DiagnosticsPresent_OutputsSkipped = 1,
         DiagnosticsPresent_OutputsGenerated = 2,
         InvalidProject_OutputsSkipped = 3,
+        ProjectReferenceCycle_OutputsSkipped = 4,
+        /** @deprecated Use ProjectReferenceCycle_OutputsSkipped instead. */
         ProjectReferenceCycle_OutputsSkupped = 4
     }
     export interface EmitResult {

--- a/tests/baselines/reference/api/typescript.d.ts
+++ b/tests/baselines/reference/api/typescript.d.ts
@@ -1964,6 +1964,8 @@ declare namespace ts {
         DiagnosticsPresent_OutputsSkipped = 1,
         DiagnosticsPresent_OutputsGenerated = 2,
         InvalidProject_OutputsSkipped = 3,
+        ProjectReferenceCycle_OutputsSkipped = 4,
+        /** @deprecated Use ProjectReferenceCycle_OutputsSkipped instead. */
         ProjectReferenceCycle_OutputsSkupped = 4
     }
     export interface EmitResult {

--- a/tests/baselines/reference/tsbuild/exitCodeOnBogusFile/initial-build/test-exit-code.js
+++ b/tests/baselines/reference/tsbuild/exitCodeOnBogusFile/initial-build/test-exit-code.js
@@ -1,0 +1,6 @@
+//// [/lib/initial-buildOutput.txt]
+/lib/tsc -b bogus.json
+error TS6053: File '/bogus.json' not found.
+exitCode:: 1
+
+

